fix(S3): Use original folder size during copy

artonge opened the pull request:
This prevents having copied folders with a wrongly set size of 0KB.

- Fix https://github.com/nextcloud/server/issues/51916

~~Not sure about the `handleCopiesAsOwned` part, but I kept it from `copyFile` to preserve the same behaviour. It was introduced in https://github.com/nextcloud/server/pull/41565. Maybe @susnux you can judge whether it makes sense or not.~~ Validated.

susnux commented on lib/private/Files/ObjectStore/ObjectStoreStorage.php:
```diff
@@ -731,6 +731,27 @@ private function copyFile(ICacheEntry $sourceEntry, string $to) {
+		$targetId = $cache->copyFromCache($cache, $sourceEntry, $to);
+
+		try {
+			if ($this->handleCopiesAsOwned) {
```
yes makes sense.
(I think the plan was to unify on one behavior, so this one should become the default (either in 32 or later)).

icewind1991 commented on lib/private/Files/ObjectStore/ObjectStoreStorage.php:
```diff
@@ -731,6 +731,27 @@ private function copyFile(ICacheEntry $sourceEntry, string $to) {
+			throw new \Exception('Invalid source cache for object store copy');
+		}
+
+		$targetId = $cache->copyFromCache($cache, $sourceEntry, $to);
```
this does a recursive copy, which would conflict with the recursion from `copyInner`

DaphneMuller commented:
@artonge anything that we should be doing here to follow up on it?

artonge commented:
> @artonge anything that we should be doing here to follow up on it?

No, just prioritizing :)

artonge committed:
fix(S3): Use original folder size during copy

This prevent having copied folders with a wrongly set size of 0KB.

- Fix https://github.com/nextcloud/server/issues/51916

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge commented:
> The least-invasive change would probably be to pass the needed metadata to `mkdir`
> 
> Alternatively you can first do the recursive cache copy, and then do the object copy recursively without changing the cache. (care has to be taken than to not have the db and object store desync if the object copy fails then)

@icewind1991 I ended up keeping the `mkdir` as is, and updating its size right after.

DaphneMuller commented:
@artonge @icewind1991 any chance this can be merged soon?

icewind1991 committed:
perf: set the folder size in the same query as we create it

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
test: add test that object store folder copy preserves folder size

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 commented:
> @icewind1991 I ended up keeping the `mkdir` as is, and updating its size right after.

Pushed a change to do the size in the `mkdir`, there is no need to do 3 queries to the filecache when 1 can suffice.

Also pushed a test to ensure this doesn't regress.
